### ISSUE TITLE
Possibly redundant branches and suboptimal codegen in anonymous functions with captured state

### ISSUE THREAD

**Issue by ancapdev** [JuliaLang/julia#22050]
_EDITED with world age workaround_

I'm working on a dataflow based event processing system where I generate state objects and event handlers at runtime. I've discovered an effect with captured state that isn't obvious to me.

If I capture a state object in an anonymous function it tends to generate reasonable code. However, it fails in some cases to propagate constants, in particular boolean state, which lead to redundant branches.

If however I call the same expression through an indirection where I pass this bound state as a parameter, codegen is able to eliminate redundant branches properly.

My code loads data from external sources through raw pointers. I haven't yet established if this is the only situation where optimization fails. In any event, here's a reproducible as minimal as I was able to make it. ```gen1()``` produces a redundant branch, whilst ```gen2()``` eliminates this branch.

Tested on ```0.5.2```, ```0.6.0-rc2```, and ```0.7.0-dev```.

```Julia
module Test

immutable Data
    a::Float32
    b::Float32
end

type State
    _3_value::Data
    _4_value::Float32
    _2_isset::Bool
    
    State() = new(Data(0, 0), 0, false)
end

const test_expr = quote
    Base.@_inline_meta
    state._2_isset = true
    if state._2_isset
        state._3_value = Data(unsafe_load(event_value, 1), unsafe_load(event_value, 2))
    end
    if state._2_isset
        state._4_value = state._3_value.a
    end
    nothing
end

function gen1(state::State)
    if (VERSION >= v"0.6-")
        Base.invokelatest(eval(:((state) -> (event_value) -> $test_expr)), state)
    else
        eval(:((state) -> (event_value) -> $test_expr))(state)
    end
end

function gen2(state::State)
    if (VERSION >= v"0.6-")
        Base.invokelatest(eval(:((sss) -> (event_value) -> ((state, event_value) -> $test_expr)(sss, event_value))), state)
    else
        eval(:((sss) -> (event_value) -> ((state, event_value) -> $test_expr)(sss, event_value)))(state)
    end
end

code_native(gen1(State()), (Ptr{Float32},))
println("----------------------------------")
code_native(gen2(State()), (Ptr{Float32},))

end
```

```Assembly
	.text
Filename: In[6]
	pushq	%rbp
	movq	%rsp, %rbp
Source line: 22
	movq	(%rdi), %rax
Source line: 18
	movb	$1, 12(%rax)
Source line: 20
	movl	(%rsi), %ecx
	movl	4(%rsi), %edx
	shlq	$32, %rdx
	orq	%rdx, %rcx
	movq	%rcx, (%rax)
Source line: 22
	testb	$1, 12(%rax)
	je	L37
Source line: 23
	movl	(%rax), %ecx
	movl	%ecx, 8(%rax)
Source line: 25
L37:
	popq	%rbp
	retq
	nopw	(%rax,%rax)
----------------------------------
	.text
Filename: In[6]
	pushq	%rbp
	movq	%rsp, %rbp
Source line: 33
	movq	(%rdi), %rax
Source line: 18
	movb	$1, 12(%rax)
Source line: 20
	movl	(%rsi), %ecx
	movl	4(%rsi), %edx
	shlq	$32, %rdx
	orq	%rcx, %rdx
	movq	%rdx, (%rax)
Source line: 23
	movl	%ecx, 8(%rax)
Source line: 33
	popq	%rbp
	retq
	nop
```

**Comment by Keno** [comment]
The example code doesn't run on 0.6+ due to world age errors. Did you post an outdated version of the code?

**Comment by ancapdev** [comment]
Thanks for spotting that.

I made changes to simplify the code and remove globals as a source of the issue, and didn't retest it on 0.6+. Not sure exactly how to fix that now, so will need to look at it later. I don't quite understand why creating anonymous functions would cause this, presumably related to redefinitions, error.

**Comment by ancapdev** [comment]
To clarify my understanding, in this minimal test case, is it right that I can't eval an expression that returns a function and then call it? What's the intended workaround?

```Julia
module Test

function test()
    eval(:(() -> nothing))()
end

test()

end
```

```
MethodError: no method matching (::Test2.##1#2)()
The applicable method may be too new: running in world age 21980, while current world is 21981.
Closest candidates are:
  #1() at In[27]:4 (method too new to be called from this world context.)
```

**Comment by Keno** [comment]
Well, calling eval at function scope is generally discouraged, but just to make it work, there's the invokelatest escape hatch:
```
function test()
           Base.invokelatest(eval(:(() -> nothing)) #= , args... =#)
end
```

**Comment by ancapdev** [comment]
Thanks, I'll try that in my real code. I'm building a system that makes heavy use of data driven runtime codegen, and evaluating expressions that return functions seemed the cleanest way to bind state into generated code (rather than going through globals).

**Comment by Keno** [comment]
Well, the escape hatch is there for a reason, but with all features like it, caveat emptor. It's easy to get into situations of excessive compile times.

**Comment by ancapdev** [comment]
Will keep that in mind, thanks.

Fixed for ```0.6+``` and retested, no change. Edited code above.

**Comment by vtjnash** [comment]
```
julia> code_native(gen1(State()), (Ptr{Float32},))
        .text
; ┌ @ REPL[4]:3 within `#4`
; │┌ @ REPL[3]:3 within `macro expansion`
        movq    (%rdi), %rax
; ││┌ @ Base.jl:43 within `setproperty!`
        movb    $1, 12(%rax)
; │└└
; │┌ @ REPL[3]:5 within `macro expansion`
; ││┌ @ pointer.jl:105 within `unsafe_load`
        vmovss  (%rsi), %xmm0                   # xmm0 = mem[0],zero,zero,zero
        vmovss  4(%rsi), %xmm1                  # xmm1 = mem[0],zero,zero,zero
; ││└
; ││┌ @ Base.jl:43 within `setproperty!`
        vmovss  %xmm0, (%rax)
        vmovss  %xmm1, 4(%rax)
; │└└
; │┌ @ REPL[3]:8 within `macro expansion`
; ││┌ @ Base.jl:43 within `setproperty!`
        vmovss  %xmm0, 8(%rax)
; │└└
        retq
        nop
; └

julia> code_native(gen2(State()), (Ptr{Float32},))
        .text
; ┌ @ REPL[5]:3 within `#8`
        movq    (%rdi), %rax
; │┌ @ REPL[5]:3 within `#9`
; ││┌ @ REPL[3]:3 within `macro expansion`
; │││┌ @ Base.jl:43 within `setproperty!`
        movb    $1, 12(%rax)
; ││└└
; ││┌ @ REPL[3]:5 within `macro expansion`
; │││┌ @ pointer.jl:105 within `unsafe_load`
        vmovss  (%rsi), %xmm0                   # xmm0 = mem[0],zero,zero,zero
        vmovss  4(%rsi), %xmm1                  # xmm1 = mem[0],zero,zero,zero
; │││└
; │││┌ @ Base.jl:43 within `setproperty!`
        vmovss  %xmm0, (%rax)
        vmovss  %xmm1, 4(%rax)
; ││└└
; ││┌ @ REPL[3]:8 within `macro expansion`
; │││┌ @ Base.jl:43 within `setproperty!`
        vmovss  %xmm0, 8(%rax)
; │└└└
        retq
        nop
; └

```
